File: hdf-converters/shared/go/formatdetect.go
package testing

import (
	"bytes"
	"encoding/json"
	"encoding/xml"
)

// InputFormat represents a detected input file format.
type InputFormat string

const (
	// FormatSARIF indicates SARIF 2.1.0 JSON input.
	FormatSARIF InputFormat = "sarif"
	// FormatJUnit indicates JUnit XML input.
	FormatJUnit InputFormat = "junit"
	// FormatXCCDF indicates XCCDF 1.2 XML input.
	FormatXCCDF InputFormat = "xccdf"
	// FormatARF indicates ARF 1.1 XML input (wraps XCCDF).
	FormatARF InputFormat = "arf"
	// FormatUnknown indicates the format could not be determined.
	FormatUnknown InputFormat = "unknown"
)

// DetectFormat examines raw input bytes and returns the detected format.
// Uses structural fingerprinting — checks for characteristic top-level fields.
//
// SARIF fingerprint: JSON object with "version" string and "runs" array.
// JUnit fingerprint: XML with <testsuites> or <testsuite> root element.
func DetectFormat(input []byte) InputFormat {
	if len(input) == 0 {
		return FormatUnknown
	}

	// Find the first non-whitespace byte to determine format family
	for _, b := range input {
		switch b {
		case ' ', '\t', '\n', '\r':
			continue
		case '{':
			return detectJSON(input)
		case '<':
			return detectXML(input)
		default:
			return FormatUnknown
		}
	}
	return FormatUnknown
}

// detectJSON probes JSON input for known formats.
func detectJSON(input []byte) InputFormat {
	var probe struct {
		Schema  string          `json:"$schema"`
		Version json.RawMessage `json:"version"`
		Runs    json.RawMessage `json:"runs"`
	}
	if err := json.Unmarshal(input, &probe); err != nil {
		return FormatUnknown
	}

	if isSARIF(probe.Version, probe.Runs, probe.Schema) {
		return FormatSARIF
	}

	return FormatUnknown
}

// detectXML probes XML input for known formats by reading the root element name.
func detectXML(input []byte) InputFormat {
	decoder := xml.NewDecoder(bytes.NewReader(input))
	for {
		tok, err := decoder.Token()
		if err != nil {
			return FormatUnknown
		}
		if se, ok := tok.(xml.StartElement); ok {
			switch se.Name.Local {
			case "testsuites", "testsuite":
				return FormatJUnit
			case "Benchmark":
				return FormatXCCDF
			case "asset-report-collection":
				return FormatARF
			default:
				return FormatUnknown
			}
		}
	}
}

// isSARIF checks whether the probed fields match SARIF structure:
// - "version" is a string (e.g. "2.1.0")
// - "runs" is an array
// - optionally "$schema" contains "sarif"
func isSARIF(version, runs json.RawMessage, schema string) bool {
	// Must have both "version" and "runs"
	if len(version) == 0 || len(runs) == 0 {
		return false
	}

	// "version" must be a JSON string
	var versionStr string
	if err := json.Unmarshal(version, &versionStr); err != nil {
		return false
	}

	// "runs" must be a JSON array
	if len(runs) == 0 || runs[0] != '[' {
		return false
	}

	return true
}

File: hdf-converters/shared/typescript/formatdetect.ts
/**
 * Input format detection via structural fingerprinting.
 *
 * Examines raw input to determine which shared format converter
 * should handle it, allowing tool-specific converters to delegate.
 */

export type InputFormat = 'sarif' | 'junit' | 'xccdf' | 'arf' | 'unknown';

/**
 * Detects the format of raw input by examining structural characteristics.
 *
 * SARIF fingerprint: JSON object with "version" (string) and "runs" (array).
 * JUnit fingerprint: XML with <testsuites> or <testsuite> root element.
 */
export function detectFormat(input: string): InputFormat {
  if (!input) {
    return 'unknown';
  }

  const trimmed = input.trim();
  if (trimmed.startsWith('{')) {
    return detectJSON(trimmed);
  }
  if (trimmed.startsWith('<')) {
    return detectXML(trimmed);
  }

  return 'unknown';
}

function detectJSON(input: string): InputFormat {
  try {
    const parsed = JSON.parse(input);
    if (typeof parsed !== 'object' || parsed === null) {
      return 'unknown';
    }

    if (isSARIF(parsed)) {
      return 'sarif';
    }
  } catch {
    return 'unknown';
  }

  return 'unknown';
}

function detectXML(input: string): InputFormat {
  // Extract root element name by finding the first opening tag after XML declaration/comments.
  // Handles both prefixed (e.g. <arf:asset-report-collection>) and unprefixed elements.
  const rootMatch = input.match(/<(?!\?|!--)(?:[a-zA-Z_][\w.-]*:)?([a-zA-Z_][\w.-]*)/);
  if (!rootMatch) {
    return 'unknown';
  }

  const rootElement = rootMatch[1];
  switch (rootElement) {
    case 'testsuites':
    case 'testsuite':
      return 'junit';
    case 'Benchmark':
      return 'xccdf';
    case 'asset-report-collection':
      return 'arf';
    default:
      return 'unknown';
  }
}

function isSARIF(obj: Record<string, unknown>): boolean {
  return typeof obj.version === 'string' && Array.isArray(obj.runs);
}
